# Cycle Data Cleaning

***Declaration ***: The central idea and coding  is abstract  from Kevin mark ham youtube video seriese, Introduction to machine learning with scikit-learn video series. You can find link under resources section. 

# Data Cleaning

In [61]:
# load libraries and set styles, options
import os,csv
import numpy as np
import pandas as pd
import seaborn as sns
import warnings; warnings.simplefilter('ignore')
#from IPython.display import HTML
#HTML('<iframe src=http://www.seattle.gov/documents/departments/sdot/newmobilityprogram/appendixb.pdf width=1000 height=350></iframe>')

In [62]:
%matplotlib inline

In [81]:
# read in a CSV
df = pd.read_csv('C:/Users/mrferozi/Desktop/GitHub/Bike/cycle-share-dataset-orignal/trip.csv', index_col=0)

In [82]:
df.index.name = 'sno'
df.reset_index(inplace=True)
df.head()

sno         starttime          stoptime    bikeid  tripduration  \
0  431  10/13/2014 10:31  10/13/2014 10:48  SEA00298       985.935   
1  432  10/13/2014 10:32  10/13/2014 10:48  SEA00195       926.375   
2  433  10/13/2014 10:33  10/13/2014 10:48  SEA00486       883.831   
3  434  10/13/2014 10:34  10/13/2014 10:48  SEA00333       865.937   
4  435  10/13/2014 10:34  10/13/2014 10:49  SEA00202       923.923   

     from_station_name                                    to_station_name  \
0  2nd Ave & Spring St  Occidental Park / Occidental Ave S & S Washing...   
1  2nd Ave & Spring St  Occidental Park / Occidental Ave S & S Washing...   
2  2nd Ave & Spring St  Occidental Park / Occidental Ave S & S Washing...   
3  2nd Ave & Spring St  Occidental Park / Occidental Ave S & S Washing...   
4  2nd Ave & Spring St  Occidental Park / Occidental Ave S & S Washing...   

  from_station_id to_station_id usertype  gender  birthyear  
0          CBD-06         PS-04   Member    Male     1960.0  
1          CBD-06         PS-04   Member    Male     1970.0  
2          CBD-06         PS-04   Member  Female     1988.0  
3          CBD-06         PS-04   Member  Female     1977.0  
4          CBD-06         PS-04   Member    Male     1971.0

In [83]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 286857 entries, 0 to 286856
Data columns (total 12 columns):
sno                  286857 non-null int64
starttime            286857 non-null object
stoptime             286857 non-null object
bikeid               286857 non-null object
tripduration         286857 non-null float64
from_station_name    286857 non-null object
to_station_name      286857 non-null object
from_station_id      286857 non-null object
to_station_id        286857 non-null object
usertype             286857 non-null object
gender               181557 non-null object
birthyear            181553 non-null float64
dtypes: float64(2), int64(1), object(9)
memory usage: 26.3+ MB


Data set contain trip duration less than three minutes, which could be possible the user took bike and decide not to proceed and return back bike to dock station and data of short pass holder which does not contribute any significance in  our analysis, at this point this study decide to get rid of unwanted data and more focus on the data which contributing significance in our analysis

In [84]:
# Filter data and only consider trip duration more than 3 minute and exclude Short-Term Pass Holder from our data set
bikes = df[(df.tripduration >=300) & (df.usertype != 'Short-Term Pass Holder')]

- After applying fillter our data set shrink from 286857 to 143541 tuples

# Handle missing values

**What does "NaN" mean?**

- "NaN" is not a string, rather it's a special value: **`numpy.nan`**.
- It stands for "Not a Number" and indicates a **missing value**.
- **`read_csv`** detects missing values (by default) when reading the file, and replaces them with this special value.

In [85]:
# 'isnull' returns a DataFrame of booleans (True if missing, False if not missing)
bikes.isnull().tail()

sno  starttime  stoptime  bikeid  tripduration  from_station_name  \
286842  False      False     False   False         False              False   
286843  False      False     False   False         False              False   
286844  False      False     False   False         False              False   
286845  False      False     False   False         False              False   
286847  False      False     False   False         False              False   

        to_station_name  from_station_id  to_station_id  usertype  gender  \
286842            False            False          False     False   False   
286843            False            False          False     False   False   
286844            False            False          False     False   False   
286845            False            False          False     False   False   
286847            False            False          False     False   False   

        birthyear  
286842      False  
286843      False  
286844      False  
286845      False  
286847      False

In [86]:
# 'nonnull' returns the opposite of 'isnull' (True if not missing, False if missing)
bikes.notnull().tail()

sno  starttime  stoptime  bikeid  tripduration  from_station_name  \
286842  True       True      True    True          True               True   
286843  True       True      True    True          True               True   
286844  True       True      True    True          True               True   
286845  True       True      True    True          True               True   
286847  True       True      True    True          True               True   

        to_station_name  from_station_id  to_station_id  usertype  gender  \
286842             True             True           True      True    True   
286843             True             True           True      True    True   
286844             True             True           True      True    True   
286845             True             True           True      True    True   
286847             True             True           True      True    True   

        birthyear  
286842       True  
286843       True  
286844       True  
286845       True  
286847       True

In [87]:
# count the number of missing values in each Series
bikes.isnull().sum()

sno                  0
starttime            0
stoptime             0
bikeid               0
tripduration         0
from_station_name    0
to_station_name      0
from_station_id      0
to_station_id        0
usertype             0
gender               0
birthyear            2
dtype: int64

**handle missing values** depends on the dataset as well as the nature of analysis

In [88]:
bikes.shape

(143543, 12)

In [89]:
# if 'any' values are missing in a row, then drop that row
bikes.dropna(how='any').shape

(143541, 12)

In [90]:
bikes.birthyear.mean()

1979.3765474672741

In [91]:
# fill in missing values with a specified value
bikes['birthyear'].fillna(value='1979', inplace=True)

In [92]:
# count the number of missing values in each Series
bikes.isnull().sum()

sno                  0
starttime            0
stoptime             0
bikeid               0
tripduration         0
from_station_name    0
to_station_name      0
from_station_id      0
to_station_id        0
usertype             0
gender               0
birthyear            0
dtype: int64

# Data Transformation 

converting data Categorical to Numaric for further use

- Converting Gender Column 

In [93]:
# create the 'Sex_num' dummy variable using the 'map' method
bikes['Sex_num'] = bikes.gender.map({'Female':2, 'Male':1})

- Converting category From Station ID and To Station ID to nominal

In [94]:
bikes["from_station_id_cat"] = bikes["from_station_id"].astype('category')
bikes.dtypes 

sno                       int64
starttime                object
stoptime                 object
bikeid                   object
tripduration            float64
from_station_name        object
to_station_name          object
from_station_id          object
to_station_id            object
usertype                 object
gender                   object
birthyear                object
Sex_num                 float64
from_station_id_cat    category
dtype: object

In [95]:
bikes["from_station_id_num"] = bikes["from_station_id_cat"].cat.codes

In [96]:
bikes["to_station_id_cat"] = bikes["to_station_id"].astype('category')
bikes.dtypes 

sno                       int64
starttime                object
stoptime                 object
bikeid                   object
tripduration            float64
from_station_name        object
to_station_name          object
from_station_id          object
to_station_id            object
usertype                 object
gender                   object
birthyear                object
Sex_num                 float64
from_station_id_cat    category
from_station_id_num        int8
to_station_id_cat      category
dtype: object

In [97]:
bikes["to_station_id_num"] = bikes["to_station_id_cat"].cat.codes

#  Handlind Date and Time

For further, for our analysis we are dividing date columns in small portion. Which, could help us to understand dataset more clearly. 

In [98]:
# convert 'Time' to datetime format
bikes['starttime'] = pd.to_datetime(bikes.starttime)
bikes['stoptime'] = pd.to_datetime(bikes.stoptime)

In [99]:
bikes.dtypes

sno                             int64
starttime              datetime64[ns]
stoptime               datetime64[ns]
bikeid                         object
tripduration                  float64
from_station_name              object
to_station_name                object
from_station_id                object
to_station_id                  object
usertype                       object
gender                         object
birthyear                      object
Sex_num                       float64
from_station_id_cat          category
from_station_id_num              int8
to_station_id_cat            category
to_station_id_num                int8
dtype: object

In [100]:
bikes['Day'] = bikes.starttime.dt.weekday_name

In [101]:
# convert 'Time' to datetime format
bikes['stoptime'] = pd.to_datetime(bikes.stoptime)

In [102]:
bikes["Day_cat"] = bikes["Day"].astype('category')
bikes.dtypes 

sno                             int64
starttime              datetime64[ns]
stoptime               datetime64[ns]
bikeid                         object
tripduration                  float64
from_station_name              object
to_station_name                object
from_station_id                object
to_station_id                  object
usertype                       object
gender                         object
birthyear                      object
Sex_num                       float64
from_station_id_cat          category
from_station_id_num              int8
to_station_id_cat            category
to_station_id_num                int8
Day                            object
Day_cat                      category
dtype: object

In [103]:
bikes["Day_num"] = bikes["Day_cat"].cat.codes

In [104]:
# convenient Series attributes are now available
bikes["sthours"] = bikes.starttime.dt.hour

In [105]:
# convenient Series attributes are now available
bikes["stphours"] = bikes.stoptime.dt.hour

In [106]:
bikes['tripduration_minutes']=bikes['tripduration']/60
desred_decimals = 2    
bikes['tripduration_minutes'] = bikes['tripduration_minutes'].apply(lambda x: round(x,desred_decimals))

In [107]:
 bikes['birthyear'] = bikes['birthyear'].astype(int)

In [108]:
bikes['age'] = 2018 - bikes['birthyear']
desred_decimals = 2    
bikes['birthyear'] = bikes['birthyear'].apply(lambda x: round(x,desred_decimals))
bikes.age.head()

0    58
1    48
2    30
3    41
4    47
Name: age, dtype: int32

Some statistics about age Variable 

In [109]:
bikes.age.describe()

count    143543.000000
mean         38.623458
std          10.251359
min          19.000000
25%          31.000000
50%          36.000000
75%          44.000000
max          87.000000
Name: age, dtype: float64

In [110]:
bikes.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 143543 entries, 0 to 286847
Data columns (total 24 columns):
sno                     143543 non-null int64
starttime               143543 non-null datetime64[ns]
stoptime                143543 non-null datetime64[ns]
bikeid                  143543 non-null object
tripduration            143543 non-null float64
from_station_name       143543 non-null object
to_station_name         143543 non-null object
from_station_id         143543 non-null object
to_station_id           143543 non-null object
usertype                143543 non-null object
gender                  143543 non-null object
birthyear               143543 non-null int64
Sex_num                 140905 non-null float64
from_station_id_cat     143543 non-null category
from_station_id_num     143543 non-null int8
to_station_id_cat       143543 non-null category
to_station_id_num       143543 non-null int8
Day                     143543 non-null object
Day_cat                 143

# Removing Duplicated Tuples

In [111]:
# detect duplicate trip_id codes: True if an item is identical to a previous item
bikes.trip_id.duplicated().tail()

AttributeError: 'DataFrame' object has no attribute 'trip_id'

The above codes shows that we do not have any duplicated rows in our dataset

In [ ]:
# Breaking date column for further Analysis
bikes['bmonth'] = bikes['starttime'].dt.month

In [ ]:
# Breaking date column for further Analysis
bikes['Date'] = bikes['starttime'].dt.date

In [ ]:
bikes.dtypes

In [ ]:
bikes.Day_num.tail()

-----------------------------------------------------

# Cleaning Weather Data

In [24]:
# read in a CSV
df = pd.read_csv('C:/Users/mrferozi/Desktop/GitHub/Bike/cycle-share-dataset-orignal/weather.csv')

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 689 entries, 0 to 688
Data columns (total 21 columns):
Date                          689 non-null object
Max_Temperature_F             689 non-null int64
Mean_Temperature_F            688 non-null float64
Min_TemperatureF              689 non-null int64
Max_Dew_Point_F               689 non-null int64
MeanDew_Point_F               689 non-null int64
Min_Dewpoint_F                689 non-null int64
Max_Humidity                  689 non-null int64
Mean_Humidity                 689 non-null int64
Min_Humidity                  689 non-null int64
Max_Sea_Level_Pressure_In     689 non-null float64
Mean_Sea_Level_Pressure_In    689 non-null float64
Min_Sea_Level_Pressure_In     689 non-null float64
Max_Visibility_Miles          689 non-null int64
Mean_Visibility_Miles         689 non-null int64
Min_Visibility_Miles          689 non-null int64
Max_Wind_Speed_MPH            689 non-null int64
Mean_Wind_Speed_MPH           689 non-null int64
Max_

#  Droping unnecessary Column

In [26]:
# remove multiple columns at once
#df.drop(['Max_Humidity', 'Min_Humidity'], axis=1, inplace=True)
#df.drop(['Max_Wind_Speed_MPH', 'Max_Gust_Speed_MPH'], axis=1, inplace=True)
#df.drop(['Max_Temperature_F', 'Min_TemperatureF'], axis=1, inplace=True)
#df.drop(['Max_Visibility_Miles',  'Min_Visibility_Miles'], axis=1, inplace=True)
#df.drop(['Max_Dew_Point_F', 'MeanDew_Point_F', 'Min_Dewpoint_F'], axis=1, inplace=True)
#df.drop(['Max_Sea_Level_Pressure_In', 'Mean_Sea_Level_Pressure_In', 'Min_Sea_Level_Pressure_In'], axis=1, inplace=True)

# Converting temprature from Fahrenheit to Celsius

The definition of precipitation is any form of water - liquid or solid - falling from the sky. It includes rain, sleet, snow, hail and drizzle plus a few less common occurrences such as ice pellets, diamond dust and freezing rain.
https://www.metoffice.gov.uk/learning/learn-about-the-weather/weather-phenomena/what-is-precipitation

- Converting  temprature from Fahrenheit to Celsius

T(°C) = (T(°F) - 32) × 5/9

Example

Convert 68 degrees Fahrenheit to degrees Celsius:


T(°C) = (68°F - 32) × 5/9 = 20 °C
https://www.rapidtables.com/convert/temperature/how-fahrenheit-to-celsius.html

In [27]:
a = df.Mean_Temperature_F-32

In [28]:
df['Mean_Temperature_C'] = a * 5/9
desred_decimals = 2    
df['Mean_Temperature_C'] = df['Mean_Temperature_C'].apply(lambda x: round(x,desred_decimals))

In [29]:
# 'isnull' returns a DataFrame of booleans (True if missing, False if not missing)
df.isnull().tail()

Date  Max_Temperature_F  Mean_Temperature_F  Min_TemperatureF  \
684  False              False               False             False   
685  False              False               False             False   
686  False              False               False             False   
687  False              False               False             False   
688  False              False               False             False   

     Max_Dew_Point_F  MeanDew_Point_F  Min_Dewpoint_F  Max_Humidity  \
684            False            False           False         False   
685            False            False           False         False   
686            False            False           False         False   
687            False            False           False         False   
688            False            False           False         False   

     Mean_Humidity  Min_Humidity         ...          \
684          False         False         ...           
685          False         False         ...           
686          False         False         ...           
687          False         False         ...           
688          False         False         ...           

     Min_Sea_Level_Pressure_In  Max_Visibility_Miles  Mean_Visibility_Miles  \
684                      False                 False                  False   
685                      False                 False                  False   
686                      False                 False                  False   
687                      False                 False                  False   
688                      False                 False                  False   

     Min_Visibility_Miles  Max_Wind_Speed_MPH  Mean_Wind_Speed_MPH  \
684                 False               False                False   
685                 False               False                False   
686                 False               False                False   
687                 False               False                False   
688                 False               False                False   

     Max_Gust_Speed_MPH  Precipitation_In  Events  Mean_Temperature_C  
684                True             False   False               False  
685                True             False   False               False  
686                True             False   False               False  
687               False             False   False               False  
688                True             False   False               False  

[5 rows x 22 columns]

In [30]:
# count the number of missing values in each Series
df.isnull().sum()

Date                            0
Max_Temperature_F               0
Mean_Temperature_F              1
Min_TemperatureF                0
Max_Dew_Point_F                 0
MeanDew_Point_F                 0
Min_Dewpoint_F                  0
Max_Humidity                    0
Mean_Humidity                   0
Min_Humidity                    0
Max_Sea_Level_Pressure_In       0
Mean_Sea_Level_Pressure_In      0
Min_Sea_Level_Pressure_In       0
Max_Visibility_Miles            0
Mean_Visibility_Miles           0
Min_Visibility_Miles            0
Max_Wind_Speed_MPH              0
Mean_Wind_Speed_MPH             0
Max_Gust_Speed_MPH            185
Precipitation_In                0
Events                          0
Mean_Temperature_C              1
dtype: int64

Converting Categorical variable to nominal to fill empty tuple for Events Variables

In [31]:
df["Events_cat"] = df["Events"].astype('category')
df.dtypes 

Date                            object
Max_Temperature_F                int64
Mean_Temperature_F             float64
Min_TemperatureF                 int64
Max_Dew_Point_F                  int64
MeanDew_Point_F                  int64
Min_Dewpoint_F                   int64
Max_Humidity                     int64
Mean_Humidity                    int64
Min_Humidity                     int64
Max_Sea_Level_Pressure_In      float64
Mean_Sea_Level_Pressure_In     float64
Min_Sea_Level_Pressure_In      float64
Max_Visibility_Miles             int64
Mean_Visibility_Miles            int64
Min_Visibility_Miles             int64
Max_Wind_Speed_MPH               int64
Mean_Wind_Speed_MPH              int64
Max_Gust_Speed_MPH              object
Precipitation_In               float64
Events                          object
Mean_Temperature_C             float64
Events_cat                    category
dtype: object

In [32]:
df["Events_num"] = df["Events_cat"].cat.codes

In [33]:
df["Events_cat"].cat.codes.head().mean()

3.0

As we can see that rain is our mean Event We will fill all empty event with 3 or rain

In [34]:
# fill in missing values with a specified value
#df['Events'].fillna(value='Rain', inplace=True)

The Events_num is appera as -1 which is not right, to fix this problem we repeate above step

In [35]:
df["Events_cat"] = df["Events"].astype('category')
df["Events_num"] = df["Events_cat"].cat.codes

In [36]:
# count the number of missing values in each Series
df.isnull().sum()

Date                            0
Max_Temperature_F               0
Mean_Temperature_F              1
Min_TemperatureF                0
Max_Dew_Point_F                 0
MeanDew_Point_F                 0
Min_Dewpoint_F                  0
Max_Humidity                    0
Mean_Humidity                   0
Min_Humidity                    0
Max_Sea_Level_Pressure_In       0
Mean_Sea_Level_Pressure_In      0
Min_Sea_Level_Pressure_In       0
Max_Visibility_Miles            0
Mean_Visibility_Miles           0
Min_Visibility_Miles            0
Max_Wind_Speed_MPH              0
Mean_Wind_Speed_MPH             0
Max_Gust_Speed_MPH            185
Precipitation_In                0
Events                          0
Mean_Temperature_C              1
Events_cat                      0
Events_num                      0
dtype: int64

In [37]:
df['Max_Gust_Speed_MPH']

0       21
1       17
2       25
3        -
4        -
5        -
6       18
7        -
8       21
9       22
10      22
11       -
12      41
13      30
14       -
15       -
16      24
17       -
18       -
19       -
20       -
21      20
22       -
23       -
24      37
25       -
26       -
27      16
28       -
29      22
      ... 
659     17
660    NaN
661     21
662    NaN
663    NaN
664    NaN
665    NaN
666    NaN
667     17
668     20
669    NaN
670    NaN
671    NaN
672    NaN
673    NaN
674    NaN
675    NaN
676    NaN
677    NaN
678    NaN
679    NaN
680    NaN
681    NaN
682    NaN
683    NaN
684    NaN
685    NaN
686    NaN
687     18
688    NaN
Name: Max_Gust_Speed_MPH, Length: 689, dtype: object

In [38]:
# fill in missing values with a specified value
df['Max_Gust_Speed_MPH'].fillna(value='0', inplace=True)
df.Max_Gust_Speed_MPH.replace('-','0',inplace=True)
df['Max_Gust_Speed_MPH']

0      21
1      17
2      25
3       0
4       0
5       0
6      18
7       0
8      21
9      22
10     22
11      0
12     41
13     30
14      0
15      0
16     24
17      0
18      0
19      0
20      0
21     20
22      0
23      0
24     37
25      0
26      0
27     16
28      0
29     22
       ..
659    17
660     0
661    21
662     0
663     0
664     0
665     0
666     0
667    17
668    20
669     0
670     0
671     0
672     0
673     0
674     0
675     0
676     0
677     0
678     0
679     0
680     0
681     0
682     0
683     0
684     0
685     0
686     0
687    18
688     0
Name: Max_Gust_Speed_MPH, Length: 689, dtype: object

In [39]:
df['Max_Gust_Speed_MPH'] = df['Max_Gust_Speed_MPH'].astype(int)
df['Max_Gust_Speed_MPH'].mean()

9.252539912917271

In [40]:
df.Max_Gust_Speed_MPH.replace('0','9.252539912917271',inplace=True)

In [41]:
# confirm that the missing values were filled in
df['Events'].value_counts().head()

Sunny         361
Rain          287
Fog            16
Fog , Rain      7
Fog-Rain        6
Name: Events, dtype: int64

In [42]:
# if 'any' values are missing in a row, then drop that row
df.dropna(how='any').shape

(688, 24)

In [43]:
# count the number of missing values in each Series
df.isnull().sum()

Date                          0
Max_Temperature_F             0
Mean_Temperature_F            1
Min_TemperatureF              0
Max_Dew_Point_F               0
MeanDew_Point_F               0
Min_Dewpoint_F                0
Max_Humidity                  0
Mean_Humidity                 0
Min_Humidity                  0
Max_Sea_Level_Pressure_In     0
Mean_Sea_Level_Pressure_In    0
Min_Sea_Level_Pressure_In     0
Max_Visibility_Miles          0
Mean_Visibility_Miles         0
Min_Visibility_Miles          0
Max_Wind_Speed_MPH            0
Mean_Wind_Speed_MPH           0
Max_Gust_Speed_MPH            0
Precipitation_In              0
Events                        0
Mean_Temperature_C            1
Events_cat                    0
Events_num                    0
dtype: int64

In [44]:
df['Mean_Temperature_F'].mean()

56.5843023255814

In [45]:
# fill in missing values with a specified value
df['Mean_Temperature_F'].fillna(value='57', inplace=True)

In [46]:
# count the number of missing values in each Series
df.isnull().sum()

Date                          0
Max_Temperature_F             0
Mean_Temperature_F            0
Min_TemperatureF              0
Max_Dew_Point_F               0
MeanDew_Point_F               0
Min_Dewpoint_F                0
Max_Humidity                  0
Mean_Humidity                 0
Min_Humidity                  0
Max_Sea_Level_Pressure_In     0
Mean_Sea_Level_Pressure_In    0
Min_Sea_Level_Pressure_In     0
Max_Visibility_Miles          0
Mean_Visibility_Miles         0
Min_Visibility_Miles          0
Max_Wind_Speed_MPH            0
Mean_Wind_Speed_MPH           0
Max_Gust_Speed_MPH            0
Precipitation_In              0
Events                        0
Mean_Temperature_C            1
Events_cat                    0
Events_num                    0
dtype: int64

In [47]:
df['Mean_Temperature_C'].mean()

13.657965116279073

In [48]:
df['Mean_Temperature_C'].fillna(value='14', inplace=True)

In [49]:
# count the number of missing values in each Series
df.isnull().sum()

Date                          0
Max_Temperature_F             0
Mean_Temperature_F            0
Min_TemperatureF              0
Max_Dew_Point_F               0
MeanDew_Point_F               0
Min_Dewpoint_F                0
Max_Humidity                  0
Mean_Humidity                 0
Min_Humidity                  0
Max_Sea_Level_Pressure_In     0
Mean_Sea_Level_Pressure_In    0
Min_Sea_Level_Pressure_In     0
Max_Visibility_Miles          0
Mean_Visibility_Miles         0
Min_Visibility_Miles          0
Max_Wind_Speed_MPH            0
Mean_Wind_Speed_MPH           0
Max_Gust_Speed_MPH            0
Precipitation_In              0
Events                        0
Mean_Temperature_C            0
Events_cat                    0
Events_num                    0
dtype: int64

In [50]:
# 'isnull' returns a DataFrame of booleans (True if missing, False if not missing)
df.isnull().tail()

Date  Max_Temperature_F  Mean_Temperature_F  Min_TemperatureF  \
684  False              False               False             False   
685  False              False               False             False   
686  False              False               False             False   
687  False              False               False             False   
688  False              False               False             False   

     Max_Dew_Point_F  MeanDew_Point_F  Min_Dewpoint_F  Max_Humidity  \
684            False            False           False         False   
685            False            False           False         False   
686            False            False           False         False   
687            False            False           False         False   
688            False            False           False         False   

     Mean_Humidity  Min_Humidity     ...      Mean_Visibility_Miles  \
684          False         False     ...                      False   
685          False         False     ...                      False   
686          False         False     ...                      False   
687          False         False     ...                      False   
688          False         False     ...                      False   

     Min_Visibility_Miles  Max_Wind_Speed_MPH  Mean_Wind_Speed_MPH  \
684                 False               False                False   
685                 False               False                False   
686                 False               False                False   
687                 False               False                False   
688                 False               False                False   

     Max_Gust_Speed_MPH  Precipitation_In  Events  Mean_Temperature_C  \
684               False             False   False               False   
685               False             False   False               False   
686               False             False   False               False   
687               False             False   False               False   
688               False             False   False               False   

     Events_cat  Events_num  
684       False       False  
685       False       False  
686       False       False  
687       False       False  
688       False       False  

[5 rows x 24 columns]

# - Breaking Date columns for further analysis

In [ ]:
# convert 'Time' to datetime format
df['Date'] = pd.to_datetime(df.Date)

In [ ]:
df['month'] = df['Date'].dt.month

In [ ]:
# remove a single column (axis=1 refers to columns)
df.drop('Events_cat', axis=1, inplace=True)

In [ ]:
df['year']=df['Date'].dt.year

# Removing Duplicated Tuples

In [ ]:
# detect duplicate Date codes: True if an item is identical to a previous item
df.Date.duplicated().tail()

In [ ]:
# count the duplicate items (True becomes 1, False becomes 0)
df.Date.duplicated().sum()

The above codes shows that we do not have any duplicated rows in our dataset

In [ ]:
df.head(100)

# Filtering Data for year 2015 and 2016

In [ ]:
df['year'] = df['year'].astype(int)

# Finalizing Bike Data and narrow down for year value

In [ ]:
# convert 'Time' to datetime format
bikes['starttime'] = pd.to_datetime(bikes.starttime)

In [ ]:
# Breaking date column for further Analysis
bikes['year'] = bikes['starttime'].dt.year

In [ ]:
df.info()

In [ ]:
# convert 'Time' to datetime format
bikes['Date'] = pd.to_datetime(bikes.Date)
bikes.info()

In [ ]:
bikes.Date.tail()

In [ ]:
df.Date.tail()

In [ ]:
# df1 belong to weather dataset
df1 = df[(df.year != 2014)]
df1.tail()

In [ ]:
# df2 belong to bikes dataset
df2 = bikes[(bikes.year != 2014) ]
bikes['Date'] = pd.to_datetime(bikes.Date)
df2.head(100)

# Saving Bike data

In [ ]:
df2.to_csv('C:\\Users\\mrferozi\\Desktop\\GitHub\\Bike\\dataset\\cycle\\trip_clean.csv')

# Saving Weather Data

Weather data consist on 600 observations where bikes data is consisting over a hundred thousand of tuples which makles it hard to combine both data set for further weather analysis.  To achieve this we will, break both weather and bikes data Date columns and applying programming technique to achieve weather value for each bicycle trip.

In [ ]:
df1.isnull().sum()

In [ ]:
df1['Mean_Temperature_C'] = df1['Mean_Temperature_C'].astype(int)
df1.dtypes

In [ ]:
df2

In [ ]:
df1['Date'] = pd.to_datetime(df1.Date)

In [ ]:
df1.to_csv('C:\\Users\\mrferozi\\Desktop\\GitHub\\Bike\\dataset\\cycle\\weather_clean.csv')

## Resources
References:*From the video series: [Introduction to machine learning with scikit-learn](https://github.com/justmarkham/scikit-learn-videos)*
- scikit-learn documentation: [Cross-validation](http://scikit-learn.org/stable/modules/cross_validation.html), [Model evaluation](http://scikit-learn.org/stable/modules/model_evaluation.html)
- scikit-learn issue on GitHub: [MSE is negative when returned by cross_val_score](https://github.com/scikit-learn/scikit-learn/issues/2439)
- Section 5.1 of [An Introduction to Statistical Learning](http://www-bcf.usc.edu/~gareth/ISL/) (11 pages) and related videos: [K-fold and leave-one-out cross-validation](https://www.youtube.com/watch?v=nZAM5OXrktY) (14 minutes), [Cross-validation the right and wrong ways](https://www.youtube.com/watch?v=S06JpVoNaA0) (10 minutes)
- Scott Fortmann-Roe: [Accurately Measuring Model Prediction Error](http://scott.fortmann-roe.com/docs/MeasuringError.html)
- Machine Learning Mastery: [An Introduction to Feature Selection](http://machinelearningmastery.com/an-introduction-to-feature-selection/)
- Harvard CS109: [Cross-Validation: The Right and Wrong Way](https://github.com/cs109/content/blob/master/lec_10_cross_val.ipynb)
- Journal of Cheminformatics: [Cross-validation pitfalls when selecting and assessing regression and classification models](http://www.jcheminf.com/content/pdf/1758-2946-6-10.pdf)